src/scancode/fingerprint.py

```diff
@@ -0,0 +1,133 @@
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# When you publish or redistribute any data created with ScanCode or any ScanCode
+# derivative work, you must accompany this data with the following acknowledgment:
+#
+#  Generated with ScanCode and provided on an "AS IS" BASIS, WITHOUT WARRANTIES
+#  OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+#  ScanCode should be considered or used as legal advice. Consult an Attorney
+#  for any legal advice.
+#  ScanCode is a free software code scanning tool from nexB Inc. and others.
+#  Visit https://github.com/nexB/scancode-toolkit/ for support and download.
+
+from bitarray import bitarray
+from licensedcode.tokenize import ngrams
+import hashlib
+
+HASH_LENGTH = 128
+SHINGLE_LENGTH = 3
+
+def get_fingerprint(location, **kwargs):
+    """
+    Return a mapping of fingerprint generated for the file at `location`.
+    """
+    with open(location, 'r') as f:
+        hashable = f.read()
+
+    simhash = Simhash()
+    simhash.update(hashable)
+    result = simhash.generate_fingerprint()
+
+    return dict(fingerprint=result)
+
+
+class Simhash:
+    """
+    Fingerprint class to generate fingerprints for files used for similarity matching
+    """
+    def __init__(self):
+        self.tokens = []
+
+    def generate_fingerprint(self):
+        """
+        Return fingerprint of the file at `location`.
+        """
+        if len(self.tokens) == 0:
+            fingerprint_hex = null
+        else:
+            weighted_hash = self.get_weighted_hash()
+            fingerprint = self.process_weighted_hash(weighted_hash)
+            fingerprint_binary_str = fingerprint.to01()
+            fingerprint_hex = self.get_hex_fingerprint_from_binary(fingerprint_binary_str)
+
+        return fingerprint_hex
+
+
+    def get_weighted_hash(self):
+        """
+        Return a weighted array from the word token list.
+        """
+        result = [0] * HASH_LENGTH
+        length = len(self.tokens) - SHINGLE_LENGTH + 1
+        shingles = ngrams(self.tokens, SHINGLE_LENGTH)
+
+        if length > 0:
+            for shingle in shingles:
+                shingle = ''.join(shingle)
+                self.process_shingles(shingle, result)
+        else:
+            self.process_shingles(''.join(self.tokens), result)
+
+        return result
+
+
+    def process_weighted_hash(self, weighted_hash):
+        """
+        Return fingerprint from a weighted array as a bitarray.
+        """
+        result = bitarray()
+
+        for item in weighted_hash:
+            if item > 0:
+                result.append(1)
+            else:
+                result.append(0)
+
+        return result
+
+
+    def bitarray_from_bytes(self, b):
+        # The returned object looks something like this:
```

Comment:
You can remove this comment. I put this here just as an example of what `bitarray.frombytes()` returns.

Reply:
This could also be a class method.